### ISSUE TITLE
Class order in Confusion Matrix

### ISSUE THREAD

**Issue by Mathieu4141** [ContinualAI/avalanche#627]
# Existing behavior

When using the Confusion Matrix in avalanche the classes are sorted per class id.

Here is a CM that can be obtained in avalanche now (0 and 3 were the last classes to be learn):
<img width="313" alt="Screen Shot 2021-05-21 at 5 46 11 PM" src="https://user-images.githubusercontent.com/23531109/119164250-83725d80-ba5c-11eb-9fae-e5f3b65a53ab.png">

# Desired behavior

In a Class-Incremental Learning setting, we usually want the classes sorted by the time they were observed in training. This allow to quickly assert if there is a bias toward the newly added classes, or toward the old classes.

For eg, here are the results showed in BiC and ICaRL papers:
<img width="340" alt="BiC" src="https://user-images.githubusercontent.com/23531109/119163843-16f75e80-ba5c-11eb-8e2c-7a1460f9ee76.png">
<img width="340" alt="ICaRL" src="https://user-images.githubusercontent.com/23531109/119163852-1959b880-ba5c-11eb-852c-e345bad29a44.png">

# Possible implementations

I think that this should only be implemented for the image rendered and not the tensor also returned by the metric, since we would loose class ids on the tensor

We could:

1. Have an optional parameter `classes_order` in the `StreamConfusionMatrix` metric

```python
        if self._save_image:
            if self.classes_order:
                cm_image = self._image_creator(exp_cm[self.class_order][:, self.class_order], display_labels=self.class_order)
            else:
                cm_image = self._image_creator(exp_cm)
            metric_representation = MetricValue(
                self, metric_name, AlternativeValues(cm_image, exp_cm),
                plot_x_position)
```

2. Provide an image creator that takes the order into account

```python 
class SortedCMImageCreator:
    def __init__(self, class_order: List[int]):
        self.class_order = class_order

    def __call__(self, tensor: Tensor) -> Image:
        return default_cm_image_creator(tensor[self.class_order][:, self.class_order], display_labels=self.class_order)
```

Option 2 seems cleaner, but more difficult to discover. Option 3 that combines the 2 could be to use a `SortedCMImageCreator` class by default when some `class_order` parameter is passed to `StreamConfusionMatrix` but no `image_creator` is set ?

What do you think?

If it is ok I'd like to make a PR for this 🙂 

EDIT: I can't seem to change the label, but this is not a bug :)

**Comment by AntonioCarta** [comment]
I agree, the current behavior is hard to read and not what everyone would expect.

> I think that this should only be implemented for the image rendered and not the tensor also returned by the metric, since we would loose class ids on the tensor

I also agree with this comment.

This should be easy to implement, MetricPlugins can track when a new class is discovered and order the labels accordingly.

Wait for @AndreaCossu before starting to work on it. He is the mantainer of the evaluation module.

**Comment by AndreaCossu** [comment]
Hi @Mathieu4141 ! Thanks for your comments. You are right, the user should have the opportunity to see classes in the order they are presented during training. This should be done at "render" time, as you correctly said. 
Following your proposals, you can implement this like: add a `absolute_class_order` (default False) input parameter to the CM with which the user can choose between an "absolute" class order based on class value or a "relative" class order based on when a class is encountered at training time. I don't think we need to allow for a custom class order. That parameter will allow to instantiate the proper image creator (the sorted one by default). If the user provides a specific image creator (we can set it to `None` by default), that one will be used and the `absolute_class_order` will be ignored.

You can proceed and try to submit a PR with these changes, that would help a lot :smile: We can discuss specific details directly in the PR. Let me know if you encounter any problems :smile: 

P.S.: while this should not be problematic for the Avalanche Confusion Matrix, it could raise some issues with the W&B ones. We may need to adapt the labels when calling the W&B function by converting them appropriately with a dictionary mapping {absolute label value: order of appearance in the training flow}. @digantamisra98 what do you think?

**Comment by digantamisra98** [comment]
@AndreaCossu Yes agreed, this would require a dictionary mapping to preserve label ordering observed over the training time.

**Comment by AndreaCossu** [comment]
#632 closed this. I will open a new issue for the W&B confusion matrix
